### ISSUE TITLE
How to create operator that listens to creation of other resource's pods

### ISSUE THREAD

**Issue by HerrSubset** [nolar/kopf#907]
### Problem

Hi :wave: 

just found out about Kopf and gave it a go. The tutorial docs are amazing, good job on that!
I'm not sure how to go about what I want to do though, and if it's even possible.

I'd like to write an operator that runs DB migrations. The plan is to tell the operator to listen for the creation of pods that have a specific label. The creation of such pods should be put on hold while the operator does the following things:

1. Check if the operator already ran the migrations
2. If not, start a new job running the DB migration container and wait for it to finish
3. Allow the pod to start normally

I was trying this approach because running DB migrations via init containers results in the migrations being run in parallel.
A Kubernetes job would also be possible, but then the whole waiting-until-job-completion and making sure there's no job name clashes would have to be part of the deployment script. An operator seemed a bit "cleaner".

I was looking at `@kopf.on.event()`, but it seems like also those handlers are tied to a specific resource on the cluster?

It would be great if someone could tell me:

- if what I want to do is possible
- which parts of Kopf I'd need to look at in order to get this working

Thank you!

**Comment by brianthelion** [comment]
@HerrSubset I'm not the package maintainer, but what you want should be very do-able with the existing handler `on.validate()`. Something like the following:

```python
@kopf.on.validate("pods",  labels={'some-label': 'some-value'}, operation="CREATE")
def _do_my_migration(logger, patch, ....):
   # create a migration pod
  migrator_pod = pykube.object.Pod(...)
  migrator_pod.create()
  for event in migrator_pod.watch():
    if event["my"]["db"]["migration"] == "failed":
       raise kopf.AdmissionError("Migration failed.")
    if event["my"]["db"]["migration"] == "success":
      break
```

The main question is whether or not your migration process will take longer than HTTP timeout on the `validate` webhook. If so, then there's no Kubernetes idiom for handling your use-case synchronously AFAIK.

**Comment by nolar** [comment]
Hello.

Keep in mind that the validation/mutation admission webhooks allow for max of 30 seconds before failing (Kubernetes's limitation). The pod startup and the migrations themselves can take much longer.

I am not sure if this task is doable gracefully: you need to suspend the pod scheduling and execution, and I do not remember if there is such a feature/mode for pods.

But it is doable via hacks: when the pod appears, modify its affinity/nodeSelector to some non-existent, put the original value to annotation. Once the migration is done, restore the original value of the affinity/nodeSelector. (Assuming that these fields are editable on pods/jobs/deployments.)

---

On the non-operator based approaches:

The much easier way, I think, would be to implement a global lock in your database (as a 1-row table), which is seen to all init containers, and is itself outside of migrations. Only one init-container will be running the migration; all other containers will be unlocked when the migration is already applied.

Or, better, consider making the migration a unit of your deployment process, not of the apps.
